File: 算法/63. 不同路径 II/UniquePathsWithObstacles.java

public class UniquePathsWithObstacles {

    class Solution {
        public int uniquePathsWithObstacles(int[][] obstacleGrid) {
            //dp方程式：dp[i][j] = dp[i - 1][j] + dp[i][j - 1]
            if (obstacleGrid.length == 0 || obstacleGrid[0][0] == 1) {
                return 0;
            }
            int m = obstacleGrid.length;
            int n = obstacleGrid[0].length;
            int[][] dp = new int[m][n];
            dp[0][0] = 1;
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    if (obstacleGrid[i][j] == 1) {
                        dp[i][j] = 0;
                        continue;
                    }
                    if (i == 0 && j > 0) {
                        dp[i][j] = dp[i][j - 1];
                    } else if (i > 0 && j == 0) {
                        dp[i][j] = dp[i - 1][j];
                    } else if (i > 0 && j > 0) {
                        dp[i][j] = dp[i][j - 1] + dp[i - 1][j];
                    }
                }
            }
            return dp[m - 1][n - 1];
        }
    }

}

File: 算法/63. 不同路径 II/UniquePathsWithObstacles.go
package _3__不同路径_II

func uniquePathsWithObstacles(obstacleGrid [][]int) int {
	if len(obstacleGrid) == 0 || obstacleGrid[0][0] == 1 {
		return 0
	}
	column := len(obstacleGrid)
	row := len(obstacleGrid[0])
	dp := make([][]int, column)
	for i := 0; i < column; i++ {
		dp[i] = make([]int, row)
	}
	dp[0][0] = 1
	for i := 0; i < column; i++ {
		for j := 0; j < row; j++ {
			if obstacleGrid[i][j] == 1 {
				dp[i][j] = 0
				continue
			}
			if i == 0 && j > 0 {
				dp[i][j] = dp[i][j-1]
			} else if j == 0 && i > 0 {
				dp[i][j] = dp[i-1][j]
			} else if i > 0 && j > 0 {
				dp[i][j] = dp[i-1][j] + dp[i][j-1]
			}
		}
	}
	return dp[column-1][row-1]
}
